### ISSUE TITLE
Rewrite rev/mat/fun/multiply.hpp using matrix derivatives

### ISSUE THREAD

**Issue by rtrangucci** [stan-dev/math#323]
#### Summary:
Precomputed gradients for matrix multiplication should use matrix derivatives for speed purposes.

#### Description:
Current matrix multiplication uses dot_product under the hood to do precomputed gradients. It is speedier to do explicit matrix calculus:

```
f(A,B) = A * B
d f(A,B)/ d A = df * B'
d f(A,B)/ d B = A' * df
```

#### Expected Output:
Speedier matrix multiplication

#### Current Version:
v2.10.0

**Comment by rtrangucci** [comment]
Todo:


1. -  [x] `multiply(Eigen::Matrix<var,-1,-1> A, Eigen::Matrix<var,-1,-1> B)`
2. - [ ] `multiply(Eigen::Matrix<double,-1,-1> A, Eigen::Matrix<var,-1,-1>  B)`
3. - [ ] `multiply(Eigen::Matrix<var,-1,-1> A, Eigen::Matrix<double,-1,-1>  B)`
4. - [x] `multiply(Eigen::Matrix<var,-1,-1> A, Eigen::Matrix<var,-1, 1>  B)`
5. - [ ] `multiply(Eigen::Matrix<double,-1,-1> A, Eigen::Matrix<var,-1, 1>  B)`
6. - [ ] `multiply(Eigen::Matrix<var,-1,-1> A, Eigen::Matrix<double,-1, 1>  B)`
7. - [x] `multiply(Eigen::Matrix<var,1,-1> A, Eigen::Matrix<var,-1, -1>  B)`
8. - [ ] `multiply(Eigen::Matrix<double,1,-1> A, Eigen::Matrix<var,-1, -1>  B)`
9. - [ ] `multiply(Eigen::Matrix<var,1,-1> A, Eigen::Matrix<double,-1, -1>  B)`
10. - [ ] `multiply(Eigen::Matrix<var,1,-1> A, Eigen::Matrix<var,-1, 1>  B)`
11. - [ ] `multiply(Eigen::Matrix<double,1,-1> A, Eigen::Matrix<var,-1, 1>  B)`
12. - [ ] `multiply(Eigen::Matrix<var,1,-1> A, Eigen::Matrix<double,-1, 1>  B)`
13. - [x] `multiply(Eigen::Matrix<var,-1, 1> A, Eigen::Matrix<var,1, -1>  B)`
14. - [ ] `multiply(Eigen::Matrix<double, -1, 1> A, Eigen::Matrix<var, 1, -1>  B)`
15. - [ ] `multiply(Eigen::Matrix<var, -1, 1> A, Eigen::Matrix<double, 1, -1>  B)`

1e0ac70 addresses 1, 4, 7, 13

**Comment by bob-carpenter** [comment]
@rtrangucci:  should this be closed now that you merged #341?

**Comment by rtrangucci** [comment]
Yes, thank you!
